manager/src/main/java/org/openremote/manager/mqtt/TeltonikaMQTTHandler.java

```diff
@@ -0,0 +1,627 @@
+package org.openremote.manager.mqtt;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import io.netty.buffer.ByteBuf;
+import io.netty.handler.codec.mqtt.MqttQoS;
+import org.apache.activemq.artemis.spi.core.protocol.RemotingConnection;
+import org.keycloak.KeycloakSecurityContext;
+import org.opengis.temporal.DateAndTime;
+import org.openremote.container.timer.TimerService;
+import org.openremote.container.util.UniqueIdentifierGenerator;
+import org.openremote.manager.asset.AssetStorageService;
+import org.openremote.manager.datapoint.AssetDatapointService;
+import org.openremote.manager.security.ManagerIdentityService;
+import org.openremote.manager.security.ManagerKeycloakIdentityProvider;
+import org.openremote.model.Container;
+import org.openremote.model.asset.Asset;
+import org.openremote.model.asset.AssetFilter;
+import org.openremote.model.asset.AssetStateDuration;
+import org.openremote.model.asset.impl.CarAsset;
+import org.openremote.model.attribute.*;
+import org.openremote.model.datapoint.AssetDatapoint;
+import org.openremote.model.query.AssetQuery;
+import org.openremote.model.query.filter.AttributePredicate;
+import org.openremote.model.query.filter.NumberPredicate;
+import org.openremote.model.syslog.SyslogCategory;
+import org.openremote.model.teltonika.*;
+import org.openremote.model.value.AttributeDescriptor;
+import org.openremote.model.value.MetaItemType;
+import org.openremote.model.value.ValueType;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.sql.Timestamp;
+import java.text.MessageFormat;
+import java.util.*;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.logging.Logger;
+import java.util.stream.Collectors;
+
+import static org.openremote.manager.event.ClientEventService.CLIENT_INBOUND_QUEUE;
+import static org.openremote.manager.mqtt.DefaultMQTTHandler.prepareHeaders;
+import static org.openremote.model.syslog.SyslogCategory.API;
+import static org.openremote.model.value.MetaItemType.*;
+
+public class TeltonikaMQTTHandler extends MQTTHandler {
+
+    //TODO: Maybe move this to Models?
+    private static class TeltonikaDevice {
```

Comment:
Prefer `protected` as it easily allows others to subclass your code and make changes.